lib/libc/powerpc64/string/strcmp.S

```diff
@@ -34,164 +34,121 @@ __FBSDID("$FreeBSD$");
 ENTRY(strcmp)
 
 .Lstrcmp_loop:
-        // Load double words.
+        /* Load double words. */
         ld %r5,0(%r3)
         ld %r6,0(%r4)
 
-        // Check if double words contain different chars.
-        cmpd  %r5,%r6
-        bne .Lstrcmp_check_differences
+        /* Check if double words are different. */
+        cmpb %r0,%r5,%r6
+        nand %r0,%r0,%r0
 
-.Lstrcmp_check_zeros:
-        // There is no different chars.
-        // Search for \0 in double word.
+        /* Check if double words contain zero. */
+        xor %r8,%r8,%r8
+        cmpb %r9,%r5,%r8
 
-        // Check 1st byte.
-        lbz %r5,0(%r3)
-        cmpdi %r5,0
-        beq .Lstrcmp_found_zero_in_0
-
-        // Check 2nd byte.
-        lbz %r5,1(%r3)
-        cmpdi %r5,0
-        beq .Lstrcmp_found_zero_in_1
-
-        // Check 3rd byte.
-        lbz %r5,2(%r3)
-        cmpdi %r5,0
-        beq .Lstrcmp_found_zero_in_2
-
-        // Check 4th byte.
-        lbz %r5,3(%r3)
-        cmpdi %r5,0
-        beq .Lstrcmp_found_zero_in_3
-
-        // Check 5th byte.
-        lbz %r5,4(%r3)
-        cmpdi %r5,0
-        beq .Lstrcmp_found_zero_in_4
+        /*
+         * If double words are different or contain zero,
+         * find what byte is different or contains zero,
+         * else load next double words.
+         */
+        or. %r9,%r9,%r0
+        bne .Lstrcmp_check_zeros_differences
 
-        // Check 6th byte.
-        lbz %r5,5(%r3)
-        cmpdi %r5,0
-        beq .Lstrcmp_found_zero_in_5
-
-        // Check 7th byte.
-        lbz %r5,6(%r3)
-        cmpdi %r5,0
-        beq .Lstrcmp_found_zero_in_6
-
-        // Check 8th byte.
-        lbz %r5,7(%r3)
-        cmpdi %r5,0
-        beq .Lstrcmp_found_zero_in_7
-
-        // increment data addresses.
+        /* Increment data addresses. */
         addi  %r3,%r3,8
         addi  %r4,%r4,8
-
         b .Lstrcmp_loop
 
-.Lstrcmp_found_zero_in_0:
-        lbz %r6,0(%r4)
-        b .Lstrcmp_end
+.Lstrcmp_check_zeros_differences:
+        /* Find what byte is different or contains zero. */
 
-.Lstrcmp_found_zero_in_1:
-        lbz %r6,1(%r4)
-        b .Lstrcmp_end
+        /* Load masks. */
+        lis %r7,0xff00
+        sldi %r5,%r7,8
```

Comment:
not sure if it makes any significant difference, or if that's even true, but it *might* be faster to make all shifts relative to %r7, instead of relative to each other:
lis %r7, 0xff00
sldi %r5, %r7, 8
sldi %r6, %r7, 16
sldi %r8, %r7, 24
sldi %r10, %r7, 32

This *could* (but, again, I'm not sure if this will happen) allow the processor to parallelize them, without having to wait for one instruction to finish before starting on the next one.

Reply:
Tried that before. It is faster to execute shift operations with lower sizes.